File: CppPythonBindingViaBoost/HelloWorld/Hello.cpp
#include <boost/python/module.hpp>
#include <boost/python/def.hpp>

#include <boost/python/class.hpp>
#include <boost/python/manage_new_object.hpp>
#include <boost/python/return_value_policy.hpp>

class Bar
{
 public:
   Bar(int x) : x(x) {}
   int get_x() const { return x; }
   void set_x(int x) { this->x = x; }
 private:
   int x;
};

class Foo
{
 public:
   Foo(int x) : b(x) {}

   // Returns an internal reference
   Bar const& get_bar() const { return b; }

 private:
   Bar b;
};

using namespace boost::python;
BOOST_PYTHON_MODULE(hello)
{
    class_<Bar>("Bar", init<int>())
        .def("get_x", &Bar::get_x)
        .def("set_x", &Bar::set_x)
        ;

    class_<Foo>("Foo", init<int>())
        .def("get_bar", &Foo::get_bar
          , return_internal_reference<>())
      ;
        
}

File: CppPythonBindingViaBoost/HelloWorld/Hello.py
#!/usr/bin/env python3

from hello import *

f = Foo(3)

b1 = f.get_bar()
b2 = f.get_bar()

b1x = b1.get_x()
b2x = b2.get_x()

print("b1x = ", b1x)
print("b2x = ", b2x)

b1.set_x(42)
b2.get_x()
b1x = b1.get_x()
b2x = b2.get_x()

print("b1x = ", b1x)
print("b2x = ", b2x)